### DISCUSSION TITLE
Recursive schema that uses other schemas not working due to incompatible inputs and outputs

### DISCUSSION THREAD

**Discussion by kronodeus** [colinhacks/zod#3068] [Q&A] [answered]
Example:

```typescript
import { z } from 'zod'

const Column = z.union([
    z.object({
        type: z.literal('string'),
    }),
    z.object({
        type: z.literal('integer'),
    }),
])

const TableBase = z.object({
    name: z.string(),
    columns: z.record(Column).default({}),
})

type TableWithExtends = z.infer<typeof TableBase> & {
    extends: TableWithExtends
}

//    Error -- Type 'undefined' is not assignable to type 'Record<string, { type: "string" } | { type: "integer" }>'
//    ||||| 
//    vvvvv
const Table: z.ZodType<TableWithExtends> = TableBase.extend({
    extends: z.lazy(() => Table),
})
```

The error goes away if I change `z.infer` to `z.input` but then the output of `Table` changes to the input as well which is not what I need. It seems since `z.ZodType` assumes output will match input, this use case falls apart, but manually supplying input and output doesn't seem to work either. Could use some pointers here.

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const Column = z.union( [
    z.object( {
        type: z.literal( 'string' ),
    } ),
    z.object( {
        type: z.literal( 'integer' ),
    } ),
] )

const TableBase = z.object( {
    name: z.string(),
    columns: z.record( Column ).default( {} ),
} )

type TableWithExtendsInput = z.input<typeof TableBase> & {
    extends: TableWithExtendsInput
}
type TableWithExtendsOutput = z.output<typeof TableBase> & {
    extends: TableWithExtendsOutput
}

const Table: z.ZodType<TableWithExtendsOutput, z.ZodTypeDef, TableWithExtendsInput> = TableBase.extend( {
    extends: z.lazy( () => Table ),
} )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by kronodeus** [comment]
Thanks, that'll work! I was trying something similar but was trying to have it so that any argument to `extends` would need to be an already validated `Table` object. In essence, something like this:

```typescript
type TableWithExtendsInput = z.input<typeof TableBase> & {
    extends: TableWithExtendsOutput // Argument should be the output type
}

type TableWithExtendsOutput = z.output<typeof TableBase> & {
    extends: TableWithExtendsOutput // Also the output type
}
```

But I think this approach was wrong-headed. Any argument to this will get validated either way, it's just a slightly different DX than I had in mind for my API.

And yes I think this should move to discussions. I didn't realize this repo had a forum. Will use that next time.
